R/rstar.R

```diff
@@ -0,0 +1,197 @@
+#'
+#' @param uncertainty Logical. Indicates whether to provide a vector of R*
+#'   values representing uncertainty in the calculated value (if `TRUE`) or a
+#'   single value (if `FALSE`). Default is `TRUE.`
+#'
+#' @param method The machine learning classifier to use (must be available in
+#'   the \pkg{caret} package). The default is `"rf"`, which calls the random
+#'   forest classifier.
+#'
+#' @param hyperparameters A named list of hyperparameter settings passed to the
+#'   classifier. Default for `"rf"` classifier is `mtry=floor(sqt(nvariables))`;
+#'   default for gradient-based model (`method="gbm"`) is `interaction.depth=3,
+#'   n.trees = 50, shrinkage=0.1, n.minobsinnode=10`.
+#'
+#' @param nsimulations Number of R* values in the returned vector if
+#'   `uncertainty` is `TRUE`. Default is 1000.
+#'
+#' @param training_proportion Proportion of iterations used to train the
+#'   classifier. Default is 0.7.
+#'
+#' @param ... Other arguments passed to `caret::train()`.
+#'
+#' @details The `rstar()` function provides a measure of MCMC convergence based
+#'   on whether it is possible to determine the chain that generated a
+#'   particular draw with a probability greater than chance. To do so, it fits a
+#'   machine learning classifier to a subset of the original MCMC draws (the
+#'   training set) and evaluates its predictive accuracy on the remaining draws
+#'   (the testing set). If predictive accuracy exceeds chance (i.e. predicting
+#'   the chain that generated a draw uniformly at random), the diagnostic
+#'   measure R* will be above 1, indicating that convergence has yet to occur.
+#'   This statistic is recently developed, and it is currently unclear what is a
+#'   reasonable threshold for diagnosing convergence.
+#'
+#'   The statistic, R*, is stochastic, meaning that each time the test is run,
+#'   unless the random seed is fixed, it will generally produce a different
+#'   result. To minimise the implications of this stochasticity, it is
+#'   recommended to repeatedly run this function to calculate a distribution of
+#'   R*; alternatively, an approximation to this distribution can be obtained by
+#'   setting `uncertainty = TRUE`, although this approximation of uncertainty
+#'   will generally have a lower mean.
+#'
+#'   By default, a random forest classifier is used (`method="rf"`), which tends
+#'   to perform best for target distributions of around 4 dimensions and above.
+#'   For lower dimensional targets, gradient boosted models (called via
+#'   `method="gbm"`) tend to have a higher classification accuracy. On a given
+#'   MCMC sample, it is recommended to try both of these classifiers.
+#'
+#' @return A numeric vector of length 1 (by default) or length `nsimulations`
+#'   (if `uncertainty=TRUE`).
+#'
+#' @references Ben Lambert, Aki Vehtari (2020) R*: A robust MCMC convergence
+#'   diagnostic with uncertainty using gradient-boosted machines \emph{arXiv
+#'   preprint} \code{arXiv:TBD}.
```

Comment:
When do you plan on submitting to arXiv? If this isn't on arXiv yet (I'm assuming since it says TBD) is there somewhere else people can find the draft? I'm a bit hesitant to add a reference to a paper until it's available to read somewhere.

Reply:
Good spot. I just added the arxiv link.